test/workers/zync_worker_test.rb

```diff
@@ -44,10 +44,140 @@ class ZyncWorkerTest < ActiveSupport::TestCase
+    end
+  end
+
+  class UnprocessableEntityRetryTest < ActiveSupport::TestCase
+    disable_transactional_fixtures!
+
+    attr_reader :worker, :event, :application
+
+    setup do
+      EventStore::Repository.stubs(raise_errors: true)
+      ZyncWorker::MessageBusPublisher.any_instance.stubs(enabled?: false)
+
+      @worker = ZyncWorker.new
+      @application = FactoryBot.create(:simple_cinstance).reload # reload to get tenant_id
+      FactoryBot.create(:simple_admin, account: application.provider_account) # for the access token
+      @event = ZyncEvent.create(RailsEventStore::Event.new, application)
+
+      worker.config.stubs(endpoint: 'http://example.com') # so it makes http request
+      worker.publisher.call(event) # so it is later available to find
+
+      stub_request(:put, 'http://example.com/tenant').to_return(status: 200)
+      stub_request(:put, 'http://example.com/notification').to_return(status: 422).with(body: event.data.to_json) # causes UnprocessableEntityError to be raised
+    end
+
+    test 'sync event dependencies when it fails' do
+      worker.expects(:sync_dependencies).with(event.event_id).returns(true)
+      worker.perform(event.event_id, event.data.with_indifferent_access)
+    end
+
+    test 'raises the error when there is no dependency' do
+      worker.expects(:sync_dependencies).with(event.event_id).returns([])
       assert_raises ZyncWorker::UnprocessableEntityError do
         worker.perform(event.event_id, event.data.with_indifferent_access)
       end
     end
+
+    test '#create_dependency_events' do
+      EventStore::Repository.expects(:find_event!).returns(event)
+      event.expects(:create_dependencies)
+      worker.create_dependency_events(event.event_id)
+    end
+
+    test '#sync_dependencies publishes events and enqueues jobs' do
+      event_id = event.event_id
+      dependency_events = event.create_dependencies
+
+      worker.expects(:create_dependency_events).with(event_id).returns(dependency_events)
+      worker.expects(:publish_dependency_events).with(dependency_events)
+
+      worker.sync_dependencies(event_id)
+    end
+
+    test '#publish_dependency_events enqueues the jobs' do
+      event_id = event.event_id
+      dependency_events = event.create_dependencies
+
+      %w[Proxy Service].each { |dependent_type| ZyncWorker.expects(:perform_async).with(anything, has_entry(:type, dependent_type)) }
+      worker.publish_dependency_events(dependency_events)
+    end
+
+    test 'on batch complete' do
+      event_id = event.event_id
+      klass = worker.class
+      klass.expects(:perform_async).with(event_id, anything, 1)
+      klass.new.on_complete(1, {'event_id' => event_id, 'manual_retry_count' => 0})
+    end
+
+    test 'number of attempts' do
+      event_id = event.event_id
+      retry_limit = worker.retry_limit
+
+      dependency_events = []
+      retry_limit.times { |i| dependency_events << event.create_dependencies }
```

Comment:
```suggestion
      retry_limit.times { dependency_events << event.create_dependencies }
```
I know that this happened for a refactoring because I am looking commit by commit 😄